src/common/kmeans.cc

```diff
@@ -0,0 +1,422 @@
+//  Copyright (C) 2019-2020 Zilliz. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License"); you may not
+//  use this file except in compliance with the License. You may obtain a copy
+//  of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+//  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+//  License for the specific language governing permissions and limitations
+//  under the License
+#include "knowhere/kmeans.h"
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <cstring>
+#include <iostream>
+#include <memory>
+#include <random>
+#include <stdexcept>
+#include <unordered_set>
+#include <vector>
+
+#include "knowhere/comp/thread_pool.h"
+#include "knowhere/comp/time_recorder.h"
+#include "knowhere/dataset.h"
+#include "knowhere/log.h"
+#include "knowhere/operands.h"
+#include "knowhere/utils.h"
+#include "simd/hook.h"
+
+namespace knowhere::kmeans {
+
+template <typename VecT>
+void
+KMeans<VecT>::elkan_L2(const VecT* x, const VecT* y, size_t d, size_t nx, size_t ny, uint32_t* ids, float* val) {
```

Comment:
my concert with elkan that it was MUCH slower that a regular implementatoin in my past experiments, up to 10x slower if I am not mistaken. Faiss uses the following BLAS call: https://github.com/facebookresearch/faiss/blob/dafdff110489db7587b169a0afee8470f220d295/faiss/utils/distances.cpp#L263

Would you consider providing a plain BLAS-based implementation as well?